### ISSUE TITLE
typings/objection - Integrating ObjectionJS with TypeScript

### ISSUE THREAD

**Issue by mastraub** [Vincit/objection.js#1548]
Coming from ten years of *Ruby on Rails* development, I'm currently working on establishing a comparable tech stack based on JavaScript/TypeScript. After losing confidence in TypeORM I'm currently putting my hope in ObjectionJS, because what I've seen so far, it gives me the impression of being a sophisticated piece of software. But still, now I'm laying my hands on, I'm having issues. Probably caused by lack of understanding, of course, but on the other hand I may have encountered issues, others may be interested in.

While the _integration test_ works fine with getting the corresponding `ValidationError` on calling `.insert(...)`, my approach for the following _unit test_ is having issues:

```TypeScript
class User extends Model {
   // ... some other code ...
   email: string;

  static get jsonSchema(): object {
    return {
      required: ['email']
    }
  }
}

// **pseudo code** I would like to have it work kinda like that,
// just that you get the idea of what I'm trying to establish:
const user = new User();
user.email = undefined; // redundant, but for sake of explicity
expect(user.$validate()).toThrow(new ValidationError('email: is a required property'));
user.email = 'a@b.c';
expect(user.$validate()).not.toThrowError();

```

At the moment I wonder why the signature of `$validate` was changed from `$validate(): void` to `$validate(json: Pojo, ...` [with this commit](https://github.com/Vincit/objection.js/commit/1a5e53e05d42a5ce37fc04d52cacd502ddb571c2).

Since `$validate` is an instance method, what kind of `json: Pojo` is supposed to be passed in here?

I would have expected something like

```TypeScript
export class Model {
  $validate(options?: ModelOptions): void;
  // or
  static validate(model: Model, options?: ModelOptions): void;
}
```
instead of 
```TypeScript
  $validate(json: Pojo, opt: ModelOptions): Pojo; // may throw ValidationError if validation fails
```

---

... so far so good ... meanwhile spending effort getting a better grasp on ObjectionJS ...

**Comment by koskimas** [comment]
The current typings have a lot of issues and they have been completely rewritten in the 2.0 branch (will be released soon). I'm not sure if this particular problem has been fixed yet, but I'll make sure it will be.

Both arguments to `$validate` method are optional, but apparently that's not reflected in the typings. In the meanwhile, you should be able to call it like this `$validate(user)`.

**Comment by mastraub** [comment]
Thanks a lot for your quick response and that you'll take care!

I checked the [typings v2.0](https://github.com/Vincit/objection.js/blob/v2.0/typings/objection/index.d.ts) as well, but the `$validate` signature is still untouched (conditions today), just so you know. I'm looking forward to v2 nevertheless.

Just to make things clear, the current v1 typed call has to look like
```TypeScript
const user = new User();
user.$validate(user, undefined);
```

**Comment by koskimas** [comment]
The reason nobody has brought this up, is that you should rarely need to call the `$validate` method explicitly. A more "objection way" of doing this is to use the static `fromJson` method, which automatically validates the input.

```js
const user = User.fromJson({ email })
```

**Comment by mastraub** [comment]
Of course, no offense intented at all. I just wanna be explicit.
Thanks for the `.fromJson` hint, works like a charm.

**Comment by koskimas** [comment]
Fixed in 2.0 branch.
